sourcegraph/sourcegraph
# cody: new config option to add custom headers

Close https://github.com/sourcegraph/sourcegraph/issues/50540
RE: https://sourcegraph.slack.com/archives/C04MSD3DP5L/p1681257445275659

This PR introduces a new config option to add custom headers to all requests made to private sg instance.

This allows users to add custom headers to all requests if a specific header is required to connect to a private Sourcegraph instance. The same method is used in our Sourcegraph VS Code Extension as requested by the same customer: https://github.com/sourcegraph/sourcegraph/issues/30916#issuecomment-1045191624

## Test plan



Result from `pnpm test:unit`:

```
 PASS   cody  src/configuration.test.ts

Test Suites: 3 passed, 3 total
Tests:       5 passed, 5 total
Snapshots:   0 total
Time:        1.919 s
Ran all test suites.
```

To test:

1. clone and pull the latest commit from this branch
2. run pnpm install from root
3. Select Launch Cody Extension from the dropdown menu in the RUN AND DEBUG sidebar in VS Code
4. Add the `"cody.customHeaders"` setting shown below to your VS Code User Setting to see the expected behavior
5. Reload VS Code
6. Ask Cody a question

Adding the following setting will not break the requests:
```
  "cody.customHeaders": {
    "Cache-Control": "no-cache",
    "Proxy-Authenticate": "Basic"
  }
  ```

Addingh the following setting will break the request to confirm the config option is being applied properly:
```
  "cody.customHeaders": {
    "Host": "www.example.com"
  }
```

<img width="1350" alt="Screenshot 2023-04-12 at 1 26 06 PM" src="https://user-images.githubusercontent.com/68532117/231576511-00864c4e-66a4-4a17-b4d6-919e23424bce.png">


## App preview:

- [Web](https://sg-web-bee-cody-header.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
